### PR TITLE
feat(ui): add better message when user doesn't have permissions to create an API key

### PR DESCRIPTION
**Pull request by y-lakhdar**
## Proposed changes

If the user logs to an organization to which he doesn't have the privileges to create an impersonate API key, he will get an Access Denied Error at the project creation.

The goal is to add a precondition to all ui:create:* project to fix 2 things:
1. Prevent the user from going through the entire project set up to finally hit an `Access Denied` error at the end of the flow.
2. Provide more information on what is preventing the user from creating a new UI project.

## Before Merging
- [ ] Merge https://github.com/coveo/platform-client/pull/367
- [ ] Remove platform-client patch

## Testing

- [x] Unit Tests:

- [ ] Functionnal Tests:

- [x] Manual Tests:

### CODE REVIEW

File: packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+export function hasNecessaryCoveoPrivileges() {
+  return async function (target: Command) {
+    const authenticatedClient = new AuthenticatedClient();
+
+    const client = await authenticatedClient.getClient();
+    const {organization} = await getConfiguration(target);
+
+    if (!(await hasCreateApiKeyPrivilege(client, organization))) {
+      // TODO: better message
+      target.warn('You cannot create an API Key');
```
Comment (y-lakhdar) [inline packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts:34]: Any idea of the message we could display?

File: packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      return false;
+    }
+    if (!(await hasImpersonatePrivilege(client, organization))) {
+      // TODO: better message
+      target.warn('You cannot create an API Key with impersonate privilege');
```
Comment (y-lakhdar) [inline packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts:39]: Same here

File: packages/cli/src/commands/ui/create/angular.spec.ts [hunk 3]
```diff
@@ -19,6 +20,7 @@ import {Config, Configuration} from '../../../lib/config/config';
 import {
   IsNpmVersionInRange,
   IsNodeVersionInRange,
+  hasNecessaryCoveoPrivileges,
```
Comment (louis-bompart) [inline packages/cli/src/commands/ui/create/angular.spec.ts:23]: Either uniformize the other preconditions in term of naming (e.g. `IsNpmVersionInRange`->`isNpmVersionInRange`) or follow what's in place ;)

File: packages/cli/src/commands/ui/create/angular.spec.ts [hunk 4]
```diff
@@ -116,13 +128,26 @@ describe('ui:create:angular', () => {
...
     preconditionStatus.node = true;
+    preconditionStatus.apiKey = true;
   });
 
   afterEach(() => {
     mockedIsNodeVersionInRange.mockClear();
     mockedIsNpmVersionInRange.mockClear();
   });
 
+  test
```
Comment (louis-bompart) [inline packages/cli/src/commands/ui/create/angular.spec.ts:139]: That's come from a nice sentiment, but I think it wont be necessary, if you check for the current preconditions, we check either if one is wrong or if all are OK.

I don't think you want or need more necessarly, you're better doing rigorous UT on the precondition function itself ;)

File: packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.spec.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+  beforeEach(() => {
+    global.config = {configDir: 'the_config_dir'} as IConfig;
+  });
+
+  afterAll(() => {
+    jest.clearAllMocks();
+  });
+
+  describe.each([
```
Comment (louis-bompart) [inline packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.spec.ts:40]: Very clean :+1:

File: packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts (hunk 1 above)
Comment (louis-bompart) [reply packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts:34]: You could infer from the `anonymous` setting to see if the user has authenticated using an access-token or the API-Key.
Like this, you can do an actionnable to the user:
 - If he's OAuthed, tell him to go bother its Coveo Admin.
 - If he's API-Key-authed, prolly he created the API Key and derped, so tell him to modify it or submit another one

File: packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts (hunk 2 above)
Comment (louis-bompart) [reply packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts:39]: Same too :troll:

**Review by louis-bompart — COMMENTED** [review]
Some minor changes, lgtm overall :gg:

File: packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+};
+
+export function HasNecessaryCoveoPrivileges() {
+  return async function (target: Command) {
+    const authenticatedClient = new AuthenticatedClient();
+
+    const client = await authenticatedClient.getClient();
+    const {organization, anonymous} = await getConfiguration(target);
+
+    console.log('*********************');
```
Comment (olamothe) [inline packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts:33]: Why those logs here ?

File: packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    console.log(anonymous);
+    console.log('*********************');
+
+    if (!(await hasCreateApiKeyPrivilege(client, organization))) {
+      target.warn(
+        anonymous
+          ? 'Your API key is missing the privilege to create other API keys. Make sure to grant this privilege before running the command again.'
+          : 'You are not authorized to create an API Key. Please contact an administrator of your Coveo organization.'
```
Comment (olamothe) [inline packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts:41]: Could we link to some article on docs.coveo.com that explains the concept better ?

File: packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+      );
+      return false;
+    }
+    if (!(await hasImpersonatePrivilege(client, organization))) {
+      target.warn(
+        anonymous
+          ? dedent`Your API key is missing the Impersonate privilege. Make sure to grant this privilege to your API key before running the command again.
+                   More info here https://docs.coveo.com/en/1707/#impersonate-domain-1`
+          : 'You are not authorized to create an API Key with the impersonate privilege. Please contact an administrator of your Coveo organization.'
```
Comment (olamothe) [inline packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts:50]: Same as above with a docs.coveo.com link.

The message for Missing impersonate privilege is 👍

**Review by olamothe — APPROVED** [review]
💯 Nice improvement !

File: packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts (hunk 6 above)
Comment (y-lakhdar) [reply packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts:33]: oups! I was doing some testing and forgot to remove it.

File: packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    const {organization, anonymous} = await getConfiguration(target);
+
+    console.log('*********************');
+    console.log(anonymous);
+    console.log('*********************');
+
+    if (!(await hasCreateApiKeyPrivilege(client, organization))) {
+      target.warn(
+        anonymous
+          ? 'Your API key is missing the privilege to create other API keys. Make sure to grant this privilege before running the command again.'
```
Comment (jpmarceau) [inline packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts:40]:
```suggestion
          ? 'Your API key is missing the privilege to create other API keys. Make sure to grant this privilege before running the command again. See https://docs.coveo.com/en/1707/#api-keys-domain.'
```

File: packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+          : 'You are not authorized to create an API Key. Please contact an administrator of your Coveo organization.'
+      );
+      return false;
+    }
+    if (!(await hasImpersonatePrivilege(client, organization))) {
+      target.warn(
+        anonymous
+          ? dedent`Your API key is missing the Impersonate privilege. Make sure to grant this privilege to your API key before running the command again.
+                   More info here https://docs.coveo.com/en/1707/#impersonate-domain-1`
```
Comment (jpmarceau) [inline packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts:49]:
```suggestion
                   See https://docs.coveo.com/en/1707/#impersonate-domain-1.`
```

File: packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts (hunk 7 above)
Comment (jpmarceau) [inline packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts:41]:
```suggestion
          : 'You are not authorized to create an API Key. Please contact an administrator of your Coveo organization and ask for that privilege. See https://docs.coveo.com/en/1707/#api-keys-domain.'
```

File: packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts (hunk 8 above)
Comment (jpmarceau) [inline packages/cli/src/lib/decorators/preconditions/apiKeyPrivilege.ts:50]:
```suggestion
          : 'You are not authorized to create an API Key with the Impersonate privilege. Please contact an administrator of your Coveo organization and ask for that privilege.  See https://docs.coveo.com/en/1707/#impersonate-domain-1.'
```

**Review by jpmarceau — APPROVED** [review]
👍

**Pull request merged**